org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/managers/InvisibleProjectBuildSupportTest.java

```diff
@@ -508,5 +508,33 @@ public void testDynamicSourceLookups() throws Exception {
 		assertNotNull(hover);
 		String javadoc = hover.getContents().getLeft().get(1).getLeft();
 		assertTrue("Unexpected Javadoc:" + javadoc, javadoc.contains("The class that manages converting HTML to Markdown"));
+
+		// add jsoup artifact to lib folder
+		File jsoupFile = DependencyUtil.getArtifact("org.jsoup", "jsoup", "1.9.2", null);
+		IPath jsoupCopy = projectRealPath.append("lib/jsoup.jar");
+		FileUtils.copyFile(jsoupFile, jsoupCopy.toFile());
+
+		// update classpath
+		UpdateClasspathJob.getInstance().updateClasspath(JavaCore.create(project), libraries);
```

Comment:
what you need to do is check if existing classpathentries already have sources attached, and retain them when rebuilding the classpath

Reply:
this is orthogonal to source discovery